rfc-144-consistent-cache-expiry.md

```diff
@@ -0,0 +1,186 @@
+or merging the frontend apps and retiring Static.
+Some frontend apps already use Content Store's max-age directive so this is
+the most straightforward way to achieve a consistent TTL on www.gov.uk pages.
+
+[random]: https://github.com/alphagov/frontend/blob/2ecb332e1e9ddabd865cb8dadb42f70e9453694e/app/controllers/random_controller.rb#L20
+[info-pages]: https://github.com/alphagov/info-frontend/blob/fad9a539f7b933c8a13a4d915de17d486c2beb7b/app/controllers/application_controller.rb#L18
+[content-store-max-age]: https://github.com/alphagov/content-store/blob/a4ea0b2b29dec4d39423ecc6b62d257b37031662/app/controllers/content_items_controller.rb#L117-L131
+
+**Removing cache invalidation logic from Emergency Banner Jenkins job**
+
+Once the TTL on www.gov.uk pages is reduced to 5 minutes it will not be necessary
+to purge the frontend application, CDN, or Varnish caches.
+
+Therefore we can remove the cache-invalidation jobs from the Emergency Banner
+Jenkins job.
+
+Unfortunately Static's template cache does not automatically expire.
+This requires us to run `rm -rf /public/templates` on all `frontend` machines
+when updating the emergency banner.
+To fix this we can have Static automatically expire the template cache every
+5 minutes as implemented in https://github.com/alphagov/static/pull/2659.
+
+## Consequences of proposals
+
+### Pros
+
+1. It will be simpler to explain how caching works: caches expire within
+  5 minutes; the banner will appear on www.gov.uk pages within 10 minutes.
+2. We won't need to maintain or debug cache invalidation behaviour
+
+### Cons
+
+1. It will take at least 5 minutes for the emergency banner to appear on www.gov.uk
+1. For 5 minutes after deployment the emergency banner will appear on some pages
+  but not on others as the objects in the CDN cache gradually expire.
+1. Invalidating caches may become harder
+1. Lower cache hit rate: At worst Origin may see a 6x increase in traffic.
+  The increase is likely to be substantially lower than this, given most traffic
+  is to extremely popular pages such as `/universal-credit`, and `/coronavirus`.
+  We will monitor the effects on performance when rolling these changes out.
+
+As caches expire, the banner will gradually appear on pages within a 10 minute window.
+The "staggered" rollout of the banner is acceptable, but could be improved upon
+if this becomes a requirement.
+If we want the banner to appear immediately on all pages, we should not use
+cache invalidation to achieve this.
+Cache invalidation is hard to get right, particularly when the objects we wish
+to invalidate exist in four layers of caching (and more than a dozen caches),
+as we have seen in incidents related to cache invalidation.
+
+Rather than using cache invalidation to immediately display a banner on all pages
+we should implement this feature in another way.
+For example, using Javascript (like a Breaking News banner) or an [edge-side include].
```

Comment:
As discussing how this can be done is out of scope of this RFC, should the examples be mentioned at all?

---

For context, the JavaScript use in the breaking news banner is not desirable - from memory, this was a quick and temporary tweak to a dismissable banner done this way because it needed to be done super speedily. It's not a good example of how things could be done when there's time.

I'd hope that any future site-wide banners that need to appear immediately wouldn't only use JavaScript - whilst it'd only exclude a low percentage of visitors, that'd still be a very large number of people.